### PR TITLE
RFC: Fapi preview fw and ima logging cel

### PR DESCRIPTION
**Pull request by JuergenReppSIT**
FAPI: Add event logging for firmware and IMA events.

 Add serialization and deserialization for firmware events:

 * Serialization and deserialization of firmware events specified in the
   TCG PC Client Platform Firmware Profile was added.
 * Also for the firmare legacy format (sha1) the serialization and deserialization
   was added.

 Add serialization and deserialization for IMA events:

 * Serialization of IMA events as described on: https://sourceforge.net/p/linux-ima/wiki/Home/
   was added. The templates ima, img-sig, ima-ng, and ima-sig are supported.
 * Tests in fapi-jons.c were adapted.

 Integrate firmware and IMA eventlog:

 * The IMA event logging was integrated.
 * The PC-Client firmware logging was integrated.

CEL Events:

* The cel_version event was added as first event for the event log of PCR 0.
* The firmware_end event was added behind the last firmware event.
* Serialization and deserialization of CEL events was added.

Event logging:

* For invalid digests an error will only produced for quote verification.
  In other cases only a warning will be displayed.
* rename "type" to "content_type".
* rename "sub_event" to "content".

Tests for event logging for firmware and IMA events:

* An Unit test for several binary examples was added.
* Tests in fapi-jons.c were adapted.
* Copied functions were removed from unit test fapi-json.c
* An Integration test for The PC-Client firmware logging was added.
* An Integration test for The PC-Client firmware logging in the
  legacy format was added.

Signed-off-by: Juergen Repp <juergen.repp@sit.fraunhofer.de>

### CODE REVIEW

File: configure.ac
Comment (AndreasFuchsTPM) [inline configure.ac:495]: IMHO this should not be turned of globally but locally with a #pragma

Comment (williamcroberts) [reply configure.ac:495]: I would ask why? Also pragma's are less portable. Ie if someone has to take the source and add a build system, that's one thing they can carry out of tree easily, but modifying the source to support pragma means carrying an out of tree inter-source file patch.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_eventlog_system.c:67]: i and itoken appear to never need to hold negative values, so wouldn't unsigned be a better choice? this way if this code gets attacked an attacker can not do negative offsets.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_eventlog_system.c:115]: generally convention is `i++` unless you need that behavior explicitly.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_eventlog_system.c:130]: is this supposed to keep going and not return false immediately?

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_eventlog_system.c:559]: This can return null, granted it requires jso to be NULL if I grok the libjson code properly:
  - https://github.com/json-c/json-c/blob/9749b0cb66b32ddac4bbbca28dd9009968722a9f/json_object.c#L1281

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_eventlog_system.c:562]: Why have get_number return an i64 when out is bounded on UINT32? Why not just use strtol and get a uint32_t?

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_eventlog_system.c:738]: extra new line to delete

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:71]: insert `SAFE_FREE(hex_string);` here and you can avoid the goto and just return if error.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:115]: memcmp so we can take advantage of quicker instructions than a byte for byte compare.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:181]: use: `src/util/aux_util.h:#define UNUSED(x) (void)(x)`

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:252]: UNUSED macro here for all 3 (void) marks.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:267]: UNUSED macro here as well

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:289]: Unused macro here as well.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:285]: I just noticed this now, inconsistent spacing on `*`. Use it tight to the variable everywhere. `UINT8 *buffer`. Here and throughout any others I missed. This seems not explicitly covered in the coding style doc, but its by and large the convention in this code base.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:333]: is the caller responsible for freeing jso?

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:355]: Won't all these assumptions break on SM3? As SM3 IIRC is 256 bits?

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:511]: shouldn't these 2 return_if_error be goto_if_error to free copy_of_template_format_orig?

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:587]: Why? Comment these magic numbers.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:717]: how come we don't `fclose(fp)` here?

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_ima_eventlog.c:825]: extra blank newline

File: src/tss2-fapi/ifapi_json_deserialize.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_json_deserialize.c:1182]: extra newline or was that intentional as a visual break between functions for the static array data?

File: src/tss2-fapi/ifapi_json_serialize.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_json_serialize.c:992]: so caller frees jso on error?

File: src/tss2-fapi/ifapi_json_serialize.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_json_serialize.c:1060]: caller frees jso on error?

File: src/tss2-fapi/ifapi_json_serialize.c
Comment (williamcroberts) [inline src/tss2-fapi/ifapi_json_serialize.c:1089]: caller frees jso on error?

**Review by williamcroberts — CHANGES_REQUESTED** [review]
i'd like to see fuzz tests to catch issues with malicious log data as inputs.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (JuergenReppSIT) [reply src/tss2-fapi/ifapi_eventlog_system.c:45]: No there is no special reason. if then would still be needed because of the second parameter of sscanf. And I would say the error handling of strtoll is more complicated.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (JuergenReppSIT) [reply src/tss2-fapi/ifapi_eventlog_system.c:130]: I will add return false.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (williamcroberts) [reply src/tss2-fapi/ifapi_eventlog_system.c:45]: > No there is no special reason. if then would still be needed because of the second parameter of sscanf. And I would say the error handling of strtoll is more complicated.

Yeah thats fine, I agree it would just be the if starts with 0x then increment itoken to hold 2 and then call scanf.

File: src/tss2-fapi/ifapi_eventlog_system.c
Comment (JuergenReppSIT) [reply src/tss2-fapi/ifapi_eventlog_system.c:562]: I did always use int64_t because we did serialize unit32 with  json_object_new_int64.

File: src/tss2-fapi/ifapi_ima_eventlog.c
Comment (JuergenReppSIT) [reply src/tss2-fapi/ifapi_ima_eventlog.c:355]: sm3 is not possible for the template-hash it should be sha1. The function is prepared for cases where e.g. sha256 will be allowed.

**Comment by JuergenReppSIT** [comment]
> i'd like to see fuzz tests to catch issues with malicious log data as inputs.
@williamcroberts The afl fuzzer revealed some bugs in the IMA part. I will test also the system event part and try to add the afl fuzzing to the ci.

**Comment by AndreasFuchsTPM** [comment]
@williamcroberts @JuergenReppSIT Let's have the discussion here:
Should we roll a new major release (4.0) because we change the CEL format or do we consider the old format was still beta anyways and not much used and just roll 3.3 after this one ?

**Comment by JuergenReppSIT** [comment]
> @williamcroberts @JuergenReppSIT Let's have the discussion here: Should we roll a new major release (4.0) because we change the CEL format or do we consider the old format was still beta anyways and not much used and just roll 3.3 after this one ?

@williamcroberts @AndreasFuchsTPM I think the new CEL format can be part of 3.3 because the old format was beta without IMA and without system events. An old log file could be converted to the new format with a simple sed script if really needed.

**Comment by williamcroberts** [comment]
> > @williamcroberts @JuergenReppSIT Let's have the discussion here: Should we roll a new major release (4.0) because we change the CEL format or do we consider the old format was still beta anyways and not much used and just roll 3.3 after this one ?
> 
> @williamcroberts @AndreasFuchsTPM I think the new CEL format can be part of 3.3 because the old format was beta without IMA and without system events. An old log file could be converted to the new format with a simple sed script if really needed.

Yeah this was what @JuergenReppSIT and I discussed before and we landed on this approach. If we can prevent a major release I am for it.

**Comment by JuergenReppSIT** [comment]
@williamcroberts @AndreasFuchsTPM AFL fuzzing has now been running for more than 5 days on 20 CPUs without finding more errors. So could the PR now be merged for 3.3?
@williamcroberts I could not get into the maintainer sync today. Was it canceled?

**Comment by williamcroberts** [comment]
@JuergenReppSIT 
- [ ] Make sure all added, non-generated, files have the BSD-2-Clause SPDX identifier
- [ ] for the bin files, could we make a build rule that generates them from b64 encoded to avoid sticking a bunch of binary blobs in the repo?
- [ ] Fix the conflict in Makefile.am

Then I think we can merge this, right?

**Comment by JuergenReppSIT** [comment]
> @JuergenReppSIT
> 
>     * [x]  Make sure all added, non-generated, files have the BSD-2-Clause SPDX identifier
> 
I have added a new commit to the PR which adds the SPDX identifier to all files without one or a wrong identifier.
>     * [ ]  for the bin files, could we make a build rule that generates them from b64 encoded to avoid sticking a bunch of binary blobs in the repo?
Is this really necessary? The tpm-tools also store binaries under test/integration/fixtures
> 
>     * [x]  Fix the conflict in Makefile.am
> 
> 
> Then I think we can merge this, right?
I would say yes.

**Comment by JuergenReppSIT** [comment]
> We should never be stuffing binary files into git. If we have to frequently update the files it can/may consume a lot of space in the .git db. I guess in this scenario it's unlikely, but I think its best to avoid any issue. We can then just have the Makefile-test.am include them as a dependency and build them using b64decode.

OK I will add it.

**Comment by williamcroberts** [comment]
> > We should never be stuffing binary files into git. If we have to frequently update the files it can/may consume a lot of space in the .git db. I guess in this scenario it's unlikely, but I think its best to avoid any issue. We can then just have the Makefile-test.am include them as a dependency and build them using b64decode.
> 
> OK I will add it.

Thanks, I know it's a pain.

**Comment by JuergenReppSIT** [comment]
> Thanks, I know it's a pain.

I think next Friday will be the day of pain for me ;-)

**Comment by JuergenReppSIT** [comment]
@williamcroberts I have added a commit where the fapi binary test files are created from base64 encoded files.

**Review by williamcroberts — APPROVED** [review]

**Pull request merged**
